Fix ABI for Int128

davidwrighton opened the pull request:
- Add ABI tests for Int128 covering interesting scenarios

Fixes #72206
(This is WIP as this first PR run is intended only to find out what problems actually exist on our various systems, following that, I will actually write the fix.)

davidwrighton committed:
Add ABI tests for Int128 covering interesting scenarios

MichalPetryka commented on src/tests/Interop/PInvoke/Int128/Int128Native.cpp:
```diff
@@ -19,6 +19,12 @@
 
+struct StructWithInt128
+{
+    int64_t messUpPadding;
```
Maybe use `uint8_t` to better validate 32bit platforms that expect 4 byte alignment instead of 8/16 ones?

davidwrighton replied:
it doesn't matter too much. The way alignment works in the various architectures we support/are likely to support, there won't be a meaningful difference. However, this is altogether messy enough that we may simply remove the ability to pass Int128 across the p/invoke boundary.

davidwrighton committed:
New test

davidwrighton committed:
Fix auto layout algorithm to compute structure alignment correctly
- In particular:
  1. The alignment requirement imposed by of a non-primitive, non-enum valuetype field is the alignment of that field
  2. The alignment requirement imposed by a primitive is the pointer size of the target platform, unless running on Arm32, in which case if the primitive or enum is 8 bytes in size, the alignment requirement is 8.

- The previous implementation produced an alignment of pointer size, unless running on Arm32 and one of the fields had an alignment requirement of 8 (in which case the alignment requirement computed for the structure would be 8)

In addition, add a test which verifies that the instance field layout test types are actually producing R2R compatible results at runtime.
 - This test shows that we have some issues around explicit layout, so I was forced to disable that portion of the test for now.

Fixes #65281

davidwrighton committed:
Re-enable disabled test

davidwrighton committed:
Remove file that shouldn't be added as part of the new test

trylek committed:
Make a few test types public to silence unassigned field errors

davidwrighton committed:
Merge branch 'fix_65281' of github.com:davidwrighton/runtime into fix_65281

davidwrighton committed:
Update comments and add more testing

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into fix_72206

davidwrighton committed:
First stab at support for proper 128bit integer layout and abi

davidwrighton committed:
Merge branch 'fix_72206' of github.com:davidwrighton/runtime into fix_72206

davidwrighton committed:
Fix bugs so that at least Windows Arm64 works

davidwrighton commented:
Closing, as this PR isn't really ready for checkin any time soon.
